src/runner.rs

```diff
@@ -124,6 +124,9 @@ impl<E: Environment> Runner<E> {
             start.elapsed().as_millis()
         );
 
+        // stop this env
+        self.env.stop(&env, db).await;
```

Comment:
This `stop` won't be called if any error happens before this line. Maybe we should ensure that this `stop` will always be executed after L92 the `start` is called.

Reply:
> This `stop` won't be called if any error happens before this line. Maybe we should ensure that this `stop` will always be executed after L92 the `start` is called.

Makes sense. Maybe I should put it into a guardian.

Reply:
I move `start()` and `stop()` out of `run_env()` in 1e456a9256db287d1f057cdfed52be09b2cdc4e5, which should prevent `stop()` from ignoring by early return. Please check it out!